### ISSUE TITLE
Strange/incorrect symtabAPI upgradePlaceholder code

### ISSUE THREAD

**Issue by jrmadsen** [dyninst/dyninst#1169]
https://github.com/dyninst/dyninst/blob/94252b8ce87293da4b10d0cee550e7ad0507a3ae/symtabAPI/src/Type-mem.h#L47

This function is causing me problems and this assert statement doesn't make sense. Assuming from the variable name `placeholder` and the usage of realloc, which explictly states:

> It must be previously allocated by malloc(), calloc() or realloc() and not yet freed with a call to free or realloc. Otherwise, the results are undefined. 

Furthermore, if you [run the example realloc code here](https://en.cppreference.com/w/c/memory/realloc), you will see that realloc _will_ change the pointer. So I am assuming that `placeholder` is already a pre-allocated memory area of `Type::max_size` so you really only need:

```cpp
template <class T>
T*
upgradePlaceholder(Type* placeholder, T* new_type)
{
    assert(sizeof(T) <= Type::max_size);
    memset(static_cast<void*>(placeholder), 0, Type::max_size);
    T* ret = new(placeholder) T{};
    assert(static_cast<void*>(placeholder) == static_cast<void*>(ret));
    *ret = *new_type;
    return ret;
}
```

In other words, since realloc requires an existing memory allocation and you are asserting that `sizeof(T) <= Type::max_size`, you don't need to realloc, you just need to do a placement new into the existing allocation.

**Comment by jrmadsen** [comment]
If I'm missing something, let me know but I'm going to open a PR for it.

**Comment by hainest** [comment]
Oh, this ol' chestnut. I discovered this a long time ago, but I never got around to fixing it because the real fix has the potential to break user code in horrible ways. Of course, the existing code can do the same thing as it's 100% undefined behavior.

The fix in #1170 was what I was going to do, initially, but decided against as it was just pushing the technical debt around. Since you are actually experiencing those asserts, then we should go ahead and at least fix that part.

**Comment by mxz297** [comment]
@hainest @jrmadsen I remember that the intention of the realloc is to save memory in place while avoiding moving memory. I agree that the current code is wrong, so it should be fixed. The problem here is that the general `Type` object is often larger than the more specific types (the `T` template), so `sizeof(Type) >= sizeof(T)` as shown by the first assertion. 

#1170 is one way to fix the wrong behavior without reclaiming the memory. The other way is to free `placeholder` and then just malloc a `T` object. I don't have any actual data to show how much memory will be wasted by merging #1170; I don't have any data to show how slower will a `free`-`malloc` sequence cause either.

Is there a way to correctly reclaiming memory in place?

**Comment by jrmadsen** [comment]
> #1170 is one way to fix the wrong behavior without reclaiming the memory.

No, #1170 reclaims the memory. A placement new does not allocate new memory.

> Is there a way to correctly reclaiming memory in place?

Yes, [placement new](https://en.cppreference.com/w/cpp/language/new#Placement_new). That is it's purpose: constructing objects at a specific (pre-allocated) memory location. It is the backbone of how memory pools work.

> The other way is to free placeholder and then just malloc a T object.

You never malloc an "object". Malloc does not call the constructor. Malloc/calloc/realloc/free are a C functions, they have no concept of constructors and destructors. `new` calls malloc to create the memory allocation and then calls the constructor. placement new skips calling malloc and calls the constructor for the object at the given memory allocation.

> I don't have any actual data to show how much memory will be wasted by merging #1170;

Zero. It doesn't change the behavior at all. `realloc` is basically a way to say: create a memory allocation starting at this address, and if it doesn't fit, deallocate the old one, and create a new allocation. Both the existing impl and #1170 will leak memory if you use the same placeholder twice without an explicit call to `~T` (destructor) before the second usage. `realloc` doesn't call the destructor because it is a C function so its usage doesn't make sense. If you are reusing these placeholders and want to avoid leaking memory, you need to do something like this:

```cpp
namespace
{
// thread-local map of callbacks to destroy a previous placement new. Visible to 
// all template instantiations of upgradePlaceholder. If re-using placeholders on
// different threads is possible, you need to remove the thread-locality and 
// add mutex locking
thread_local auto destructor_cb = std::unordered_map<Type*, void (*)(Type*)>{};
}

template<class T>
T *upgradePlaceholder(Type *placeholder, T *new_type)
{
    assert(sizeof(T) <= Type::max_size);

    // see if there is a destructor callback and, if so, call it
    auto itr = destructor_cb.find(placeholder);
    if(itr != destructor_cb.end()) { itr->second(placeholder); }

    memset(static_cast<void*>(placeholder), 0, Type::max_size);
    // callback ensures that placement new isn't overwriting existing object
    T* ret = new(placeholder) T{};
    assert(static_cast<void*>(placeholder) == static_cast<void*>(ret));

    // create a new callback to call the destructor
    destructor_cb[placeholder] = [](Type* p) { static_cast<T*>(p)->~T(); };

    *ret = *new_type;
    return ret;
}
```

**Comment by jrmadsen** [comment]
Ignore that last part of the comment with respect to `destructor_cb`. I didn't realize `T` was always derived from `Type` so a manual "virtual table" of destructors like this is unnecessary.
